File: TSP-NearestNeighbour.cpp
#include <stdio.h>
#include <iostream>
#include <stack>
#include <vector>
#include<fstream>

using namespace std;
vector<vector<int> > map;
int distance1;
stack<int> stk;
int numberOfNodes;

vector<bool> used;
vector<int> name;
string s;
int n;

void input()
{
	int temp;
	ifstream inf("TSP.txt");
	inf >> n;
	for (int i = 0; i < n; i++)
	{
		map.push_back(vector<int>());
		used.push_back(false);
		name.push_back(49 + i);
		for (int j = 0; j < n; j++)
		{
			inf >> temp;
			map[i].push_back(temp);
		}
	}
	inf.close();
}

void tsp(vector<vector<int> > map) {
	numberOfNodes = n - 1;
	vector<int> visited(numberOfNodes + 1, 0);
	visited[1] = 1;
	stk.push(1);
	int element = 0;
	int dst;
	int i;
	int min = 2147483647;
	bool minFlag = false;
	cout << 1 + " ";

	while (!stk.empty())
	{
		element = stk.top();
		i = 1;
		min = 2147483647;
		while (i <= n-1)
		{
			if (map[element][i] > 1 && visited[i] == 0)
			{
				if (min > map[element][i])
				{
					min = map[element][i];
					dst = i;
					minFlag = true;
				}
			}
			i++;
		}
		if (minFlag)
		{
			distance1 += map[element][dst];
			visited[dst] = 1;
			stk.push(dst);
			cout << dst + " ";
			minFlag = false;
			continue;
		}
		stk.pop();
	}
}

int main() {
	input();
	tsp(map);
}

File: TSPNearestNeighbour.java
import java.io.FileNotFoundException;
import java.io.FileReader;
import java.util.Scanner;
import java.util.Stack;
 
public class TSPNearestNeighbour
{
    private int numberOfNodes;
    private Stack<Integer> stack;
    private static int distance;
    private static int rtn;
 
    public TSPNearestNeighbour()
    {
        stack = new Stack<Integer>();
    }
 
    public void tsp(int adjacencyMatrix[][])
    {
        numberOfNodes = adjacencyMatrix[1].length - 1;
        int[] visited = new int[numberOfNodes + 1];
        visited[1] = 1;
        stack.push(1);
        int element, dst = 0, i;
        int min = Integer.MAX_VALUE;
        boolean minFlag = false;
        System.out.print(1 + " ");
 
        while (!stack.isEmpty())
        {
            element = stack.peek();
            i = 1;
            min = Integer.MAX_VALUE;
            while (i <= numberOfNodes)
            {
                if (adjacencyMatrix[element][i] > 1 && visited[i] == 0)
                {
                    if (min > adjacencyMatrix[element][i])
                    {
                        min = adjacencyMatrix[element][i];
                        dst = i;
                        minFlag = true;
                    }
                }
                i++;
            }
            if (minFlag)
            {
            	distance += adjacencyMatrix[element][dst];
                visited[dst] = 1;
                stack.push(dst);
                System.out.print(dst + " ");
                minFlag = false;
                continue;
            }
            rtn = dst;
            stack.pop();
        }
    }
 
    public static void main(String[] arg) throws FileNotFoundException
    {	
    	long startMili=System.currentTimeMillis();
        int number_of_nodes;
        Scanner scanner = new Scanner(new FileReader("TSP.txt"));
        number_of_nodes = scanner.nextInt();
        int adjacency_matrix[][] = new int[number_of_nodes + 1][number_of_nodes + 1];
        for (int i = 1; i <= number_of_nodes; i++)
        	{
                for (int j = 1; j <= number_of_nodes; j++)
                {
                    adjacency_matrix[i][j] = scanner.nextInt();
                }
            }
        for (int i = 1; i <= number_of_nodes; i++)
            {
                for (int j = 1; j <= number_of_nodes; j++)
                {
                    if (adjacency_matrix[i][j] == 1 && adjacency_matrix[j][i] == 0)
                    {
                        adjacency_matrix[j][i] = 1;
                    }
                }
            }
        System.out.println("the citys are visited as follows");
        TSPNearestNeighbour tspNearestNeighbour = new TSPNearestNeighbour();
        tspNearestNeighbour.tsp(adjacency_matrix);
        System.out.println();
        long endMili=System.currentTimeMillis();
        distance = distance + adjacency_matrix[1][rtn];
        System.out.println("distance: " + distance);
        System.out.println("run time: " + (endMili - startMili) + "ms");
        scanner.close();
    }
}
